fdb-record-layer-core/src/main/java/com/apple/foundationdb/record/query/plan/cascades/CascadesPlanner.java

```diff
@@ -502,58 +558,48 @@ public interface Task {
+            return plannerPhase;
         }
 
         @Override
         public void execute() {
-            if (group.needsExploration()) {
-                // Explore the group, then come back here to pick an optimal expression.
-                taskStack.push(this);
-                for (RelationalExpression member : group.getMembers()) {
-                    // enqueue explore expression which then in turn enqueues necessary rules for transformations
-                    // and matching
-                    exploreExpressionAndOptimizeInputs(context, group, member, false, evaluationContext);
-                }
-                // the second time around we want to visit the else and prune the plan space
-                group.startExploration();
-            } else {
-                RelationalExpression bestMember = null;
-                for (RelationalExpression member : group.getMembers()) {
-                    if (bestMember == null || new CascadesCostModel(configuration).compare(member, bestMember) < 0) {
-                        if (bestMember != null) {
-                            // best member is being pruned
-                            traversal.removeExpression(group, bestMember);
-                        }
-                        bestMember = member;
-                    } else {
-                        // member is being pruned
-                        traversal.removeExpression(group, member);
+            RelationalExpression bestFinalExpression = null;
+            for (final var finalExpression : group.getFinalExpressions()) {
+                if (bestFinalExpression == null ||
+                        plannerPhase.createCostModel(configuration).compare(finalExpression, bestFinalExpression) < 0) {
+                    if (bestFinalExpression != null) {
+                        // best member is being pruned
+                        traversal.removeExpression(group, bestFinalExpression);
                     }
+                    bestFinalExpression = finalExpression;
+                } else {
+                    // member is being pruned
+                    traversal.removeExpression(group, finalExpression);
                 }
-                if (bestMember == null) {
-                    throw new RecordCoreException("there we no members in a group expression used by the Cascades planner");
-                }
-
-                group.pruneWith(bestMember);
-                group.commitExploration();
+            }
+            if (bestFinalExpression == null) {
+                group.clearFinalExpressions();
+            } else {
+                group.pruneWith(bestFinalExpression);
             }
         }
 
         @Override
         public Debugger.Event toTaskEvent(final Location location) {
-            return new Debugger.OptimizeGroupEvent(currentRoot, taskStack, location, group);
+            return new Debugger.OptimizeGroupEvent(plannerPhase, currentRoot, taskStack, location, group);
```

Comment:
`plannerPhase` is an explicit attribute of every `Task` as there are tasks within the same planner run that are in different phases.